app/src/main/java/org/mozilla/fenix/settings/DeleteBrowsingDataController.kt

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.settings
+
+import android.content.Context
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.engine.Engine
+import mozilla.components.feature.tab.collections.TabCollection
+import org.mozilla.fenix.FeatureFlags
+import org.mozilla.fenix.ext.components
+import kotlin.coroutines.CoroutineContext
+
+interface DeleteBrowsingDataController {
+    suspend fun deleteTabs()
+    suspend fun deleteBrowsingData()
+    suspend fun deleteCollections(collections: List<TabCollection>)
+    suspend fun deleteCookies()
+    suspend fun deleteCachedFiles()
+    suspend fun deleteSitePermissions()
+}
+
+class DefaultDeleteBrowsingDataController(
+    val context: Context,
+    val coroutineContext: CoroutineContext = Dispatchers.Main
+) : DeleteBrowsingDataController {
+    override suspend fun deleteTabs() {
+        withContext(coroutineContext) {
+            context.components.useCases.tabsUseCases.removeAllTabs.invoke()
+        }
+    }
+
+    override suspend fun deleteBrowsingData() {
+        withContext(coroutineContext) {
+            if (FeatureFlags.granularDataDeletion) {
+                context.components.core.engine.clearData(Engine.BrowsingData.select(Engine.BrowsingData.DOM_STORAGES))
+            } else {
+                context.components.core.engine.clearData(Engine.BrowsingData.all())
+            }
+        }
+        context.components.core.historyStorage.deleteEverything()
+    }
+
+    override suspend fun deleteCollections(collections: List<TabCollection>) {
+        while (context.components.core.tabCollectionStorage.getTabCollectionsCount() != collections.size) {
+            delay(DELAY_IN_MILLIS)
+        }
+
+        collections.forEach { context.components.core.tabCollectionStorage.removeCollection(it) }
+    }
+
+    override suspend fun deleteCookies() {
+        withContext(coroutineContext) {
+            context.components.core.engine.clearData(Engine.BrowsingData.select(Engine.BrowsingData.COOKIES))
+        }
+    }
+
+    override suspend fun deleteCachedFiles() {
+        withContext(coroutineContext) {
+            context.components.core.engine.clearData(
+                Engine.BrowsingData.select(Engine.BrowsingData.ALL_CACHES)
+            )
+        }
+    }
+
+    override suspend fun deleteSitePermissions() {
+        withContext(coroutineContext) {
+            context.components.core.engine.clearData(
+                Engine.BrowsingData.select(Engine.BrowsingData.ALL_SITE_SETTINGS)
```

Comment:
I think we should remove this from this category because `ALL_SITE_SETTINGS` includes `PERMISSIONS` (no-op IIUC from the GV team) and then caches which the user can clear separately?

Reply:
All site settings does not include any caches, as far as I can tell. I'm not sure if it's entirely a no-op either though.

```
            const val COOKIES: Int = 1 shl 0
            const val NETWORK_CACHE: Int = 1 shl 1
            const val IMAGE_CACHE: Int = 1 shl 2
            const val DOM_STORAGES: Int = 1 shl 4
            const val AUTH_SESSIONS: Int = 1 shl 5
            const val PERMISSIONS: Int = 1 shl 6
            const val ALL_CACHES: Int = NETWORK_CACHE + IMAGE_CACHE
            const val ALL_SITE_SETTINGS: Int = (1 shl 7) + PERMISSIONS
            const val ALL_SITE_DATA: Int = (1 shl 8) + COOKIES + DOM_STORAGES + ALL_CACHES + ALL_SITE_SETTINGS
            const val ALL: Int = 1 shl 9
```

Reply:
Ah okay I missed that `ALL_SITE_SETTINGS` didn't include `ALL_CACHES`. I guess if it's just clearing `PERMISSIONS` that may or may not do anything in addition to using `permissionStorage` this could be fine